### PR TITLE
chore: add `pkg.pr.new`

### PR DESCRIPTION
**Pull request by userquin**
### Description

This PR includes:
- ~update PR template to include `/publish` tip~
- ~add CR comment workflow via `/publish` (tip on the PR template)~
- add CR workflow to use `cr-tracked`  label PR or main branch push
- release all packages: we can add one workflow per package releasing only the corresponding one

This PR requires:
- create a `cr-tracked` label
- add [pkg-pr-new app](https://github.com/apps/pkg-pr-new) to vitest package
- ~add `CR_PAT` secret to Actions with a token with `public_repo` (Classic Token): later we can automate it with something like this https://github.com/vitest-dev/vitest/blob/main/.github/workflows/ecosystem-ci-trigger.yml#L63C1-L68 (I'm a newbye on GH actions/webflows)~

This PR requires also a review to exclude CR for `docs` for example, we can use something like this in `on.pull_request`:
```yaml
    paths-ignore:
      - '.github/**'
      - 'test/**'
      - 'examples/**'      
      - 'docs/**'
      - '*.md'
```

/cc @Aslemammad

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: .github/workflows/cr-comment.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+permissions:
+  pull-requests: write
+
+jobs:
+  label:
+    if: ${{ github.event.issue.pull_request && (github.event.comment.user.id == github.event.issue.user.id || github.event.comment.author_association == 'MEMBER' || github.event.comment.author_association == 'COLLABORATOR') && startsWith(github.event.comment.body, '/publish') }}
+    runs-on: ubuntu-latest
+
+    steps:
+      - run: gh issue edit ${{ github.event.issue.number }} --add-label cr-tracked --repo ${{ github.repository }}
```
Comment (Aslemammad) [inline .github/workflows/cr-comment.yml:16]: hmm, what is this for?

Comment (userquin) [reply .github/workflows/cr-comment.yml:16]: will add the `cr-tracked` label once receive `/publish` PR comment

Comment (Aslemammad) [reply .github/workflows/cr-comment.yml:16]: ohhh, makes sense! Thank you so much!

**Comment by Aslemammad** [comment]
I love this.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -17,3 +17,8 @@
 
 ### Changesets
 - [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.
+
+---
+
+> [!TIP]
+> The author of this PR can publish a _preview release_ by commenting `/publish` below.
```
Comment (AriPerkkio) [inline .github/PULL_REQUEST_TEMPLATE.md:24]: So can anyone, even non-collaborators, publish `vitest` packages to pkg.pr.new?

Is it tied to Vitest's `secrets.CR_PAT` at least? Meaning that does the service prevents random users from publishing `vitest` and `@vitest/*` packages to the registry?

Comment (userquin) [reply .github/PULL_REQUEST_TEMPLATE.md:24]: No, only PR author, members and collaborators, that's CR.

Comment (sheremet-va) [reply .github/PULL_REQUEST_TEMPLATE.md:24]: PR author is "anyone"

File: .github/workflows/cr.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
+name: CR
```
Comment (sheremet-va) [inline .github/workflows/cr.yml:1]: I don't understand why we need this and can't do that in the comment workflow?

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (AriPerkkio) [reply .github/PULL_REQUEST_TEMPLATE.md:24]: Looks like the URL actually has the repository owner specified too by default. So no malicious users can publish `https://pkg.pr.new/vitest@<hash>` packages, I think. The `--compact` flag probably does some checks for the origin of the release, @Aslemammad?

Comment (userquin) [reply .github/PULL_REQUEST_TEMPLATE.md:24]: > PR author is "anyone"

That's CR, we can change it to allow only members to publish via comment or cr-tracked label. There is no npm registry it is just tgz files.

File: .github/workflows/cr.yml (hunk 3 above)
Comment (userquin) [reply .github/workflows/cr.yml:1]: I'm newbie to gh actions/workflows, feel free to play with other layouts.

You can check vitepress integration, brc-cc did some changes about the publish comment, actions cannot fire workflows, maybe can be done: 
- original PR https://github.com/vuejs/vitepress/pull/406
- including cr-comment workflow https://github.com/vuejs/vitepress/commit/1d629c5c4fdb582c357d9d917a1ab0d3df738687

Comment (userquin) [reply .github/workflows/cr.yml:1]: You can check vite publish-comment workflow, I use it with no luck.

Comment (sheremet-va) [reply .github/workflows/cr.yml:1]: Ah, yes! Vite workflow is what I expected

**Comment by hi-ogawa** [comment]
@userquin We discussed about how to trigger pkg.pr.new and we agreed "cr-tracked" label based workflow is okay. I'll push a few commits here and see how it goes.

**Comment by hi-ogawa** [comment]
@sheremet-va We haven't installed the app yet? https://github.com/vitest-dev/vitest/actions/runs/11050071341/job/30697038656?pr=6362#step:7:13

```
{"url":"/check","statusCode":404,"statusMessage":"Not Found","message":"The app https://github.com/apps/pkg-pr-new is not installed on vitest-dev/vitest."}
```

**Comment by sheremet-va** [comment]
> @sheremet-va We haven't installed the app yet? [vitest-dev/vitest/actions/runs/11050071341/job/30697038656?pr=6362#step:7:13](https://github.com/vitest-dev/vitest/actions/runs/11050071341/job/30697038656?pr=6362#step:7:13)
> 
> ```
> {"url":"/check","statusCode":404,"statusMessage":"Not Found","message":"The app https://github.com/apps/pkg-pr-new is not installed on vitest-dev/vitest."}
> ```

We didn't install anything

**Comment by sheremet-va** [comment]
Should be installed now

**Comment by userquin** [comment]
you need to add the app to vitest repo: https://github.com/stackblitz-labs/pkg.pr.new?tab=readme-ov-file#setup

**Comment by userquin** [comment]
> @userquin We discussed about how to trigger pkg.pr.new and we agreed "cr-tracked" label based workflow is okay. I'll push a few commits here and see how it goes.

Check vitepress repo, using vue bot, check `cr.yml` and `cr-comment.yml` here: https://github.com/vuejs/vitepress/tree/main/.github/workflows

**Comment by hi-ogawa** [comment]
@userquin `cr-comment.yml` doesn't look necessary since we can just add label manually when we want to allow PRs to be released. Isn't it only doing `/publish` comment to trigger adding label `cr-tracked` indirectly?

**Comment by userquin** [comment]
@hi-ogawa  I've sent you an invitation to vue discord server and a link to Vitepress room/channel, I asked `brc_dd` and his answer:

> Actually, github actions can also add labels but labels added via their default bot cannot trigger other actions. 🫠 That's why add it using separate user.

**Comment by sheremet-va** [comment]
> Actually, github actions can also add labels but labels added via their default bot cannot trigger other actions. 🫠 That's why add it using separate user.

We don't want to add labels automatically. The idea is that the maintainer adds a label to allow CR manually

**Comment by hi-ogawa** [comment]
Yeah, I just also checked re-labeling "cr-tracked" and this alone can also trigger CR correctly. https://github.com/vitest-dev/vitest/actions/workflows/cr.yml

image

**Comment by hi-ogawa** [comment]
I tested a few examples locally and also on stackblitz and it looks working https://stackblitz.com/edit/vitest-dev-vitest-ei9jaw?file=package.json
I think we can proceed with this PR.

**Comment by patak-dev** [comment]
For reference, adding the label automatically refers to this scheme:
- https://github.com/vitejs/vite/pull/18211

I think adding a label instead of a comment is quite nice too, so good to move forward with this PR as is. I think in Vite we'll keep the comment (with the label workaround). We can try both approaches and see what ends up working best.

**Review by sheremet-va — COMMENTED** [review]
Do we need a mention in the readme/documentation that we have this? We have this for example: https://vitest.dev/guide/#using-unreleased-commits

**Comment by userquin** [comment]
> Do we need a mention in the readme/documentation that we have this? We have this for example: https://vitest.dev/guide/#using-unreleased-commits

https://github.com/vite-pwa/vite-plugin-pwa/blob/main/.github/pull_request_template.md

**Comment by sheremet-va** [comment]
> > Do we need a mention in the readme/documentation that we have this? We have this for example: [vitest.dev/guide#using-unreleased-commits](https://vitest.dev/guide/#using-unreleased-commits)
> 
> [vite-pwa/vite-plugin-pwa@`main`/.github/pull_request_template.md](https://github.com/vite-pwa/vite-plugin-pwa/blob/main/.github/pull_request_template.md?rgh-link-date=2024-09-30T16%3A12%3A11Z)

We also publish every commit on `main`. And the comment can only be added by the maintainers

**Comment by hi-ogawa** [comment]
I just pushed a few commits but `pkg-pr-new publish` is getting stuck today.
@Aslemammad Could it be some infra issue? Would be great to know if there's something we can do to diagnose issue.

**Comment by Aslemammad** [comment]
interesting, can you trigger a re-run for the failed workflow?

**Comment by hi-ogawa** [comment]
Thanks for the help. I started a re-run and it looks like it's still stuck.

**Comment by Aslemammad** [comment]
I think it is fixed now, just triggered a re-run!

**Comment by hi-ogawa** [comment]
@Aslemammad Thanks! I confirmed pkg-pr-new is working now https://github.com/vitest-dev/vitest/actions/runs/11115894967/job/30904769639?pr=6362

**Comment by Aslemammad** [comment]
Happy to hear that ❤️

**Pull request merged**